py/objstrunicode.c

```diff
@@ -183,28 +183,23 @@ STATIC mp_obj_t str_subscr(mp_obj_t self_in, mp_obj_t index, mp_obj_t value) {
         // load
 #if MICROPY_PY_BUILTINS_SLICE
         if (mp_obj_is_type(index, &mp_type_slice)) {
-            mp_obj_t ostart, ostop, ostep;
-            mp_obj_slice_get(index, &ostart, &ostop, &ostep);
-            if (ostep != mp_const_none && ostep != MP_OBJ_NEW_SMALL_INT(1)) {
+            mp_bound_slice_t slice;
+            mp_obj_slice_indices(index, utf8_charlen(self_data, self_len), &slice);
```

Comment:
This call to `utf8_charlen` is going to unconditionally scan the whole (unicode) string on each slice access, to compute the character length.  Note that the original code here, with the calls to `str_index_to_ptr`, was optimised to not scan for the (common) case of small positive start index and small negative end index.  Probably the best/easiest thing to do here is just not touch this part of the code.

Reply:
Yes, I spotted that. This is however a time/space trade-off. If I put it back to the old way then not only is the code larger but I need to reinstate `mp_obj_slice_get()` (which is otherwise unneeded), so we end up with quite a bit more code. Do you have hard statistics for the distribution of the offset sizes used in slices or just a hunch?

Reply:
> Do you have hard statistics for the distribution of the offset sizes used in slices or just a hunch?

No.  I was just going by the existing comments in `str_index_to_ptr`.  But the change here would mean that something like `unicode_str[:-1]` will scan the unicode bytes twice, once to determine the number of characters, and once to count len-1 characters.

Yes it's a trade off, but IMO it seems good to keep the current optimised behaviour of this unicode indexing.

Maybe `mp_obj_slice_get()` can be just inlined here instead?

Reply:
OK. I reversed out the refactoring in `str_subscr()` and then in-lined the access to the slice components.

The net result is that with the addition of support for the `.indices()` method, the exposed C API for the `indicies()` functionality, the refactored code in `sequence.c` and the in-lined slice fetch in `str_subscr()`, the whole thing is 32 bytes *smaller* than before I started on the ESP32.

The remaining question is: since this PR both implements a missing Python feature and seems to make the code smaller, should I make it unconditional?